File: app/src/test/java/com/photoviewer/domain/interactor/GetPhotoDetailsTest.java
package com.photoviewer.domain.interactor;

import com.photoviewer.data.entity.PhotoEntity;
import com.photoviewer.data.repository.PhotoEntityDataSource;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import rx.Scheduler;
import rx.Observable;

import static org.mockito.BDDMockito.given;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.verifyZeroInteractions;

public class GetPhotoDetailsTest {
    private static final int FAKE_PHOTO_ID = 123;

    private GetPhotoDetails mGetPhotoDetails;

    @Mock
    private PhotoEntityDataSource mMockPhotoEntityDataSource;
    @Mock
    private Scheduler mMockScheduler;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        mGetPhotoDetails = new GetPhotoDetails(mMockScheduler,
                mMockScheduler, mMockPhotoEntityDataSource);
    }

    @Test
    public void testGetPhotoDetails() {
        given(mMockPhotoEntityDataSource.photo(anyInt())).willReturn(Observable.just(new PhotoEntity()));

        mGetPhotoDetails.setPhotoId(FAKE_PHOTO_ID);
        mGetPhotoDetails.buildObservable();

        verify(mMockPhotoEntityDataSource).photo(FAKE_PHOTO_ID);
        verifyNoMoreInteractions(mMockPhotoEntityDataSource);
        verifyZeroInteractions(mMockScheduler);
    }
}

File: app/src/test/java/com/photoviewer/domain/usecases/GetPhotoDetailsTest.kt
@file:Suppress("IllegalIdentifier")

package com.photoviewer.domain.usecases

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import com.photoviewer.data.entity.PhotoEntity
import com.photoviewer.data.repository.PhotoEntityDataSource
import com.photoviewer.domain.Photo
import org.junit.Before
import org.junit.Test
import org.mockito.Matchers.anyInt
import rx.Observable
import rx.Scheduler
import rx.observers.TestSubscriber
import kotlin.test.assertEquals

class GetPhotoDetailsTest {

    private val mPhotoEntityDataSource: PhotoEntityDataSource = mock()
    private val mScheduler: Scheduler = mock()

    private lateinit var mGetPhotoDetails: GetPhotoDetails
    private lateinit var mTestSubscriber: TestSubscriber<Photo>


    companion object {
        private val FAKE_PHOTO_ID = 123
    }

    @Before fun setUp() {
        mTestSubscriber = TestSubscriber.create()
        mGetPhotoDetails = GetPhotoDetails(mScheduler, mScheduler, mPhotoEntityDataSource)
    }

    @Test fun `should get particular photo`() {
        val photoEntity = createPhotoEntity()
        assumeDataSourceHasRequestedPhoto(photoEntity)

        mGetPhotoDetails.setPhotoId(FAKE_PHOTO_ID)
        mGetPhotoDetails.call().subscribe(mTestSubscriber)

        assertEquals(FAKE_PHOTO_ID, mTestSubscriber.onNextEvents[0].id)
    }

    private fun createPhotoEntity() = PhotoEntity().apply {
        id = FAKE_PHOTO_ID
    }

    private fun assumeDataSourceHasRequestedPhoto(photoEntity: PhotoEntity) {
        whenever(mPhotoEntityDataSource.photo(anyInt())).thenReturn(
                Observable.just(photoEntity))
    }
}
